sdk/dotnet/Pulumi/X/Automation/PulumiFn.cs

```diff
@@ -0,0 +1,9 @@
+﻿using System.Collections.Generic;
+
+namespace Pulumi.X.Automation
+{
+    /// <summary>
+    /// A Pulumi program as an inline function (in process).
+    /// </summary>
+    public delegate IDictionary<string, object?> PulumiFn();
```

Comment:
I do wonder if we should have this return `Task<IDictionary<string, object?>` instead? It would allow consumers to make their `PulumiFn` asynchronous, but it would also make synchronous functions a bit uglier because they would need to return `Task.FromResult(output)`. 

We could have a `PulumiFnAsync` and a `PulumiFn` at the same time, and just have an overload for `PulumiFn` anywhere that we take in `PulumiFnAsync`.. that might be a bit easier to swallow. But all of our options objects would have `PulumiFnAsync` and you wouldn't be able to set that with a `PulumiFn` directly.

I also think that if we do have a `PulumiFnAsync`, that it should take in a `CancellationToken` - specifically the one we get from the CLI call so that if the CLI cancels our inline function is made aware.